pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -186,6 +409,185 @@ func (s *DiscoveryServer) updateCluster(push *model.PushContext, clusterName str
+// This is the main method implementing EDS.
+// It replaces InstancesByPort in model - instead of iterating over all endpoints it uses
+// the hostname-keyed map. And it avoids the conversion from Endpoint to ServiceEntry to envoy
+// on each step: instead the conversion happens once, when an endpoint is first discovered.
+func (s *DiscoveryServer) EDSUpdate(shard, serviceName string,
+	entries []*model.IstioEndpoint) error {
+	return s.edsUpdate(shard, serviceName, entries, false)
+}
+
+// BeforePush is a callback invoked just before the push. It currently swaps and returns the
+// EdsUpdates map - additional preparation will be added as we move to full incremental.
+// This is needed to keep things isolated and use the right mutex.
+// Once proxy/envoy/discovery is merged into v2 discovery this can become non-public.
+func (s *DiscoveryServer) BeforePush() map[string]*model.ServiceShards {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	edsUpdates := s.edsUpdates
+	// Reset - any new updates will be tracked by the new map
+	s.edsUpdates = map[string]*model.ServiceShards{}
+
+	return edsUpdates
+}
+
+func (s *DiscoveryServer) edsUpdate(shard, serviceName string,
+	entries []*model.IstioEndpoint, internal bool) error {
+	// edsShardUpdate replaces a subset (shard) of endpoints, as result of an incremental
+	// update. The endpoint updates may be grouped by K8S clusters, other service registries
+	// or by deployment. Multiple updates are debounced, to avoid too frequent pushes.
+	// After debounce, the services are merged and pushed.
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	// Update the data structures for the service.
+	// 1. Find the 'per service' data
+	ep, f := s.EndpointShardsByService[serviceName]
+	if !f {
+		// This endpoint is for a service that was not previously loaded.
+		// Return an error to force a full sync, which will also cause the
+		// EndpointsShardsByService to be initialized with all services.
+		ep = &model.ServiceShards{
+			Shards:          map[string]*model.EndpointShard{},
+			ServiceAccounts: map[string]bool{},
+		}
+		s.EndpointShardsByService[serviceName] = ep
+		if !internal {
+			adsLog.Infof("Full push, new service %s", serviceName)
+			s.ConfigUpdater.ConfigUpdate(true)
+		}
+	}
+
+	// 2. Update data for the specific cluster. Each cluster gets independent
+	// updates containing the full list of endpoints for the service in that cluster.
+	ce := &model.EndpointShard{
+		Shard:   shard,
+		Entries: []*model.IstioEndpoint{},
+	}
+
+	for _, e := range entries {
+		ce.Entries = append(ce.Entries, e)
+		if e.ServiceAccount != "" {
+			_, f = ep.ServiceAccounts[e.ServiceAccount]
+			if !f && !internal {
+				// The entry has a service account that was not previously associated.
+				// Requires a CDS push and full sync.
+				adsLog.Infof("Endpoint updating service account %s %s", e.ServiceAccount, serviceName)
+				s.ConfigUpdater.ConfigUpdate(true)
```

Comment:
rather than doing a full on push, can we simply trigger a CDS cache recompute and do a CDS push to all ? Unfortunately, it will cause corresponding EDS updates as well, which is very wasteful

Reply:
Not in this PR. I spent 4 weeks on it ( with all the label and service account fixing ), I have other things to do.

Reply:
Besides - it would be wasted once incremental XDS is available.